botnim/benchmark/assistant_loop.py

```diff
@@ -64,34 +81,190 @@ def assistant_loop(client: OpenAI, assistant_id, question=None, thread=None, not
 
-        if run.status == 'completed': 
+        if run.status == 'completed':
+            # Log assistant's response when run is completed
+            messages = client.beta.threads.messages.list(thread_id=thread.id, order='desc', limit=1)
+            for message in messages:
+                if message.role == "assistant":
+                    with open(log_file, 'a', encoding='utf-8') as f:
+                        f.write("\nAssistant Response:\n")
+                        for content in message.content:
+                            if content.type == 'text':
+                                f.write(f"{content.text.value}\n")
+                                break  # Only write the latest response
+            
+            # Log completion
+            with open(log_file, 'a', encoding='utf-8') as f:
+                f.write("\n=== Run Completed ===\n\n")
             break
 
         for tool in run.required_action.submit_tool_outputs.tool_calls:
             arguments = json.loads(tool.function.arguments)
+            output = None  # Initialize output variable
+            
+            # Log tool input
+            with open(log_file, 'a', encoding='utf-8') as f:
+                f.write(f"\nTool Input:\n  Tool ID: {tool.id}\n  Name: {tool.function.name}\n  Arguments:\n")
+                for key, value in arguments.items():
+                    f.write(f"    {key}: {value}\n")
+            
+            # Set default page_size for DatasetDBQuery
             if tool.function.name == 'DatasetDBQuery':
                 arguments['page_size'] = 30
-            output = get_openapi_output(openapi_spec, tool.function.name, arguments)
-            if tool.function.name == 'DatasetInfo':
-                output = get_dataset_info_cache(arguments, output)
-            tool_outputs.append(dict(
-                tool_call_id=tool.id,
-                output=json.dumps(output, ensure_ascii=False, indent=2)
-            ))
-        run = client.beta.threads.runs.submit_tool_outputs_and_poll(
-            thread_id=thread.id,
-            run_id=run.id,
-            tool_outputs=tool_outputs
-        )
-        assert run.status in ['completed', 'requires_action'], f'RUN STATUS: {run.status}'    
+                
+            # Handle different tool types
+            if tool.function.name.startswith('search_'):
+                # Handle the search_takanon__context__dev pattern
+                # Remove 'search_' prefix and '__dev' suffix if present
+                tool_name = tool.function.name[len('search_'):]
+                if tool_name.endswith('__dev'):
+                    tool_name = tool_name[:-len('__dev')]
+                
+                # Split into bot_name and context_name
+                parts = tool_name.split('__', 1)
+                bot_name = parts[0]
+                context_name = parts[1] if len(parts) > 1 else ''
+                
+                # Load config to get context settings
```

Comment:
Why do we need this bit of code before running query?

Reply:
to handle the general case where different tools might reference different bots. I know we are not using it now (single bot, single context), but prefer to keep it, as this is just a local debugging tool, and I prefer not to re-write it later.

Reply:
Everything under this line and up to the `run_query` call is only used to get 'num_results'... which you already do in the `run_query` function. 
How is that related to `different tools might reference different bots`?

Reply:
sorry, I misunderstood your original comment. You are very right, fixed.